x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,520 @@
+    })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetQuotas { channel_id } => get_quotas(deps, channel_id),
+    }
+}
+
+fn get_quotas(deps: Deps, channel_id: impl Into<String>) -> StdResult<Binary> {
+    to_binary(&CHANNEL_FLOWS.load(deps.storage, &channel_id.into())?)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    unimplemented!()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use cosmwasm_std::testing::{mock_dependencies, mock_env, mock_info};
+    use cosmwasm_std::{from_binary, Addr, Attribute};
+
+    use crate::helpers::tests::verify_query_response;
+    use crate::msg::{Channel, QuotaMsg};
+    use crate::state::RESET_TIME_WEEKLY;
+
+    const IBC_ADDR: &str = "IBC_MODULE";
+    const GOV_ADDR: &str = "GOV_MODULE";
+
+    #[test]
+    fn proper_instantiation() {
+        let mut deps = mock_dependencies();
+
+        let msg = InstantiateMsg {
+            gov_module: Addr::unchecked(GOV_ADDR),
+            ibc_module: Addr::unchecked(IBC_ADDR),
+            channels: vec![],
+        };
+        let info = mock_info(IBC_ADDR, &vec![]);
+
+        // we can just call .unwrap() to assert this was a success
+        let res = instantiate(deps.as_mut(), mock_env(), info, msg).unwrap();
+        assert_eq!(0, res.messages.len());
+
+        // The ibc and gov modules are properly stored
+        assert_eq!(IBCMODULE.load(deps.as_ref().storage).unwrap(), IBC_ADDR);
+        assert_eq!(GOVMODULE.load(deps.as_ref().storage).unwrap(), GOV_ADDR);
+    }
+
+    #[test]
+    fn permissions() {
+        let mut deps = mock_dependencies();
+
+        let quota = QuotaMsg::new("Weekly", RESET_TIME_WEEKLY, 10, 10);
+        let msg = InstantiateMsg {
+            gov_module: Addr::unchecked(GOV_ADDR),
+            ibc_module: Addr::unchecked(IBC_ADDR),
+            channels: vec![Channel {
+                name: "channel".to_string(),
+                quotas: vec![quota],
+            }],
+        };
+        let info = mock_info(IBC_ADDR, &vec![]);
+        instantiate(deps.as_mut(), mock_env(), info.clone(), msg).unwrap();
+
+        let msg = ExecuteMsg::SendPacket {
+            channel_id: "channel".to_string(),
+            channel_value: 3_000,
+            funds: 300,
+        };
+
+        // This succeeds
+        execute(deps.as_mut(), mock_env(), info.clone(), msg).unwrap();
+
+        let info = mock_info("SomeoneElse", &vec![]);
+
+        let msg = ExecuteMsg::SendPacket {
+            channel_id: "channel".to_string(),
+            channel_value: 3_000,
+            funds: 300,
+        };
+        let err = execute(deps.as_mut(), mock_env(), info.clone(), msg).unwrap_err();
+        assert!(matches!(err, ContractError::Unauthorized { .. }));
+    }
+
+    #[test]
+    fn consume_allowance() {
```

Comment:
I think we should add a test ensuring that quotas roll over as expected

Reply:
This is done implicitly here: https://github.com/osmosis-labs/osmosis/blob/1f237e8fb0f04a8fec2e17812be0295cbdcdffc3/x/ibc-rate-limit/contracts/rate-limiter/src/integration_tests.rs#L64, but I could check the internal state to make sure the reset actually sets the values to zero. 

The reason to have that in integration test is because we need to modify the time, so we need a mock chain